### PR TITLE
feat: Support for binary data streaming

### PR DESCRIPTION
**Pull request by mladkau**
Adds support for streaming of binary data. Using the `data` and `end` events the request response on the client side is switched to streaming and each received `chunk` is send over the websocket. The server forwards each received `chunk` directly to the consuming service.

### CODE REVIEW

File: lib/client/socket.js [hunk 1]
```diff
@@ -30,10 +30,12 @@ module.exports = ({ url, token, filters, config, identifyingMetadata }) => {
 
   logger.info({ url }, 'broker client is connecting to broker server');
 
-  const response = relay.response(filters, config);
+  const response = relay.response(filters, config, io);
+  const streamingResponse = relay.streamingResponse;
 
   // RS note: this bind doesn't feel right, it feels like a sloppy way of
   // getting the filters into the request function.
+  io.on('chunk', streamingResponse(token));
```
Comment (mladkau) [inline lib/client/socket.js]: We have now a new `chunk` message which contains chunks of a data stream.

File: lib/relay.js [hunk 2]
```diff
@@ -8,12 +8,41 @@ const replace = require('./replace-vars');
...
+const stream = require('stream');
 
 module.exports = {
   request: requestHandler,
   response: responseHandler,
+  streamingResponse: streamResponseHandler,
 };
 
+const binaryRequest = [
+  'application/vnd.docker.image.rootfs.diff.tar.gzip',
```
Comment (mladkau) [inline lib/relay.js]: For now only very specific `accept` headers will trigger the streaming code.

File: lib/relay.js [hunk 3]
```diff
@@ -8,12 +8,41 @@ const replace = require('./replace-vars');
...
+];
+
+const streams = new Map();
```
Comment (mladkau) [inline lib/relay.js]: Maybe there is a better way to bind the server response to the chunk handler. I do think the connection pooling gets in the way and I haven't found a way to remove event listeners from a socker i..e `off(streamResponseHandler)`

For now each response object of a binary streaming request gets a UUID which the streamResponseHandler uses to send the incoming chunks into the right response.

File: lib/relay.js [hunk 4]
```diff
@@ -43,6 +72,29 @@ function requestHandler(filterRules) {
 
       req.url = result.url;
       logContext.ioUrl = result.url;
+
+      // check if this is a streaming request for binary data
+      if (req.headers && binaryRequest.indexOf(req.headers['accept']) !== -1) {
```
Comment (mladkau) [inline lib/relay.js]: Identify the streaming request by the accept header.

File: lib/relay.js [hunk 5]
```diff
@@ -43,6 +72,29 @@ function requestHandler(filterRules) {
...
+
+        res.locals.io.send('request', {
+          url: req.url,
+          method: req.method,
+          body: req.body,
+          headers: req.headers,
+          streamingID,
+        });
+
+        return;
```
Comment (mladkau) [inline lib/relay.js]: We don't set header or anything else in the response as the whole response is streamed directly from the remote system. All filtering must have happened at this point.

File: lib/relay.js [hunk 6]
```diff
@@ -197,15 +251,31 @@ function responseHandler(filterRules, config) {
...
-      }, (error, response, body) => {
+      };
+
+      // check if this is a streaming request for binary data
+      if (streamingID) {
+        logger.debug(logContext, 'serving binary stream request');
+
+        req.encoding = null; // no encoding for binary data
+
+        request(req).on('data', chunk => {
```
Comment (mladkau) [inline lib/relay.js]: As mentioned above - we pipe all chunks of the response directly to the broker server which then forwards everything directly to the consuming service.

File: lib/relay.js (hunk 4 above)
Comment (adrukh) [reply lib/relay.js]: consider `req.get('accept')` instead?

Comment (jurglic) [reply lib/relay.js]: Is `req.headers &&` necessary?

File: lib/relay.js [hunk 7]
```diff
@@ -197,37 +250,72 @@ function responseHandler(filterRules, config) {
...
+        }).on('error', error => {
+          logError(logContext, error);
+          io.send('chunk', streamingID, error.message, true, {
+            status : 500
+          });
+        });
+        return;
+      }
+
+      request(req, (error, response, body) => {
```
Comment (mladkau) [inline lib/relay.js]: I've tried to merge the two request statements into one but this got quite messy and didn't reduce the code much. I feel that the two mechanisms are distinctive enough to warrant two branches.

File: test/fixtures/server/filters.json [hunk 8]
```diff
@@ -28,10 +28,21 @@
...
+      "method": "GET",
+      "origin": "http://localhost:${originPort}",
     }
 
   ],
   "public": [
+    {
+      "path": "/test-blob/*",
+      "method": "GET",
+      "stream": true
```
Comment (mladkau) [inline test/fixtures/server/filters.json]: Changing this flag will fail the tests as the response will be encoded by the non-streaming code.

File: lib/relay.js (hunk 3 above)
Comment (moshikod) [reply lib/relay.js]: I think this is fairly a correct approach to handle this async collection of binary chunks.
Consider adding some mechanism to limit the maximum memory consumption together with an ageing limit to handle the case where chunks streaming was stopped from some reason.

File: test/utils.js [hunk 9]
```diff
@@ -15,6 +15,27 @@ const { app: echoServer, server } = webserver({
   httpsCert: process.env.TEST_CERT, // Optional
 });
 
+echoServer.get(
+  '/test-blob/1',
+  (req, res) => {
+    res.setHeader('test-orig-url', req.originalUrl);
+    res.status(299);
```
Comment (moshikod) [inline test/utils.js]: I there any special reason for using 299?

File: lib/relay.js (hunk 3 above)
Comment (mladkau) [reply lib/relay.js]: I think the aging limit is a fair point and I'll add a TTL to the entries (using node-cache). 

In terms of memory consumption: We create a streamingID for each binary download - we would need a fair amount of downloads before we reach any kind of significant memory limit (the streaming objects are only PassThrough objects and should not consume a lot of memory). I think I would prefer to wait here and see if we run into memory issues before considering this.

**Pull request merged**